velox/functions/sparksql/String.h

```diff
@@ -1431,4 +1432,208 @@ struct LevenshteinDistanceFunction {
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        StringView{kMaskedLowerCase_},
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr,
+      const arg_type<Varchar>* otherCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        getMaskedChar(otherCharPtr));
+    return true;
+  }
+
+ private:
+  void doCall(
+      out_type<Varchar>& result,
+      StringView input,
+      const std::optional<StringView> upperChar,
+      const std::optional<StringView> lowerChar,
+      const std::optional<StringView> digitChar,
+      const std::optional<StringView> otherChar) const {
+    auto inputBuffer = input.data();
+    const size_t inputSize = input.size();
+    result.reserve(inputSize);
+    auto outputBuffer = result.data();
+    size_t inputIdx = 0;
+    size_t outputIdx = 0;
+    while (inputIdx < inputSize) {
+      utf8proc_int32_t curCodePoint;
+      int charByteSize;
+      curCodePoint = utf8proc_codepoint(
+          &inputBuffer[inputIdx], inputBuffer + inputSize, charByteSize);
+      charByteSize = curCodePoint == -1 ? 1 : charByteSize;
+      auto maskedChar = &inputBuffer[inputIdx];
+      auto maskedCharByteSize = charByteSize;
+      // Treat invalid UTF-8 character as other char.
```

Comment:
Is this the same behavior with Spark? Perhaps document this in string.rst.

Reply:
not exactlly same, spark mask function actually has issue when handling invalid UTF-8 character or wide character (4 byte). For example for wide character  "😀" which is 2 bytes `\uD83D\uDE00`, spark would wrongly treat it 2 characters, `\uD83D` and `\uDE00`. For invalided UTF-8 character case, for example `\xED`, spark would treat it as 4 characters - "\", "x", "E", "D". That's due to the limitation of spark use java `toString.map` to iterater the char which only work well for character which has at most 2 bytes (16 bits). Checking spark code - https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala#L322C1-L349C4

For do right thing, in our implement, for those wide character cases or invalid UTF-8 character cases we treat it correctlly as single character, instead of aligning with spark.

Reply:
Does Spark contain tests for invalid UTF-8 character and wide character (4 byte) like you mentioned?

Reply:
Don't have, this are the tests from spark i found - https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala#L5772C1-L5804C4  

and https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/inputs/mask-functions.sql

which we also already coverred in our tests.

Reply:
Thanks for providing above information. As you mentioned, the difference with Spark is on the handling of `invalid UTF-8 character and wide character`.

I tried in Spark and found its output was inconsistent indeed, e.g., for below case this PR gives `yyyyyいいYyYYdddいいい` while Spark gives `yyyyyいいYyYYdddいいいい` where one more `い` is added.

Is it more like undefined behaviors in Spark because there is no test coverage for those cases? And the result from this PR looks more right from my view. 

> spark.sql("select mask('синяяい绿AbCD123世界🙂', 'Y', 'y', 'd', 'い')").show(false)
+-----------------------------------------+
|mask(синяяい绿AbCD123世界🙂, Y, y, d, い)|
+-----------------------------------------+
|yyyyyいいYyYYdddいいいい                 |
+-----------------------------------------+


@mbasmanova Do you have any suggestion? Thanks.

Reply:
Presto explicitly states that string functions assume valid UTF-8 input. I assume Spark is the same, but maybe documentation is missing. Might be good to open an issue in Spark to ask whether this is the case and if so to add documentation. Then, we can add the same clause to Velox's documentation for string Spark functions.

https://prestodb.io/docs/current/functions/string.html
```
These functions assume that the input strings contain valid UTF-8 encoded Unicode code points. There are no explicit checks for valid UTF-8 and the functions may return incorrect results on invalid UTF-8.
```

Reply:
@rui-mo , as i already mentioned spark also can't also handle wide-width character correctly, 
java `toString` default using `UTF-16` if my unserstanding right. And java char class object use 16 bits array to represents a character that means the spark mask function can only handle weel for BMP char (that is 16 bits).

for `世界🙂`, where `世界` is chinese character which each one has 2 bytes, so it would be replaced correctly with `いい`, but `🙂` is a wide character which represents with 4 bytes, so spark wrongly treat it as 2 character each one have 2 bytes, causing `🙂` replacing with `いい`.

Reply:
let me know your concern @rui-mo

Reply:
> Might be good to open an issue in Spark to ask whether this is the case and if so to add documentation. Then, we can add the same clause to Velox's documentation for string Spark functions.

> https://prestodb.io/docs/current/functions/string.html

@gaoyangxiaozhu I understand your point. Perhaps we can follow this suggestion to ask about it in Spark. How do you think?

Reply:
thanks @rui-mo , I believe from the view of code implement part of spark mask function, they don't concern for `invalid UTF-8` and `wide character`.

 I just raise a issue/ question to stackoverflow https://stackoverflow.com/staging-ground/78781459 and also Spark JIRA issue - https://issues.apache.org/jira/browse/SPARK-48973

let's waitting the spark community's reply.

Reply:
Thanks!